[LLD][AArch64][ARM] Delay adding long thunk mapping symbols

smithp35 opened the pull request:
When we create a thunk we don't know whether it will be short or long. Move the emission of the long thunk mapping symbol to when we transition to a long thunk. This improves disassembly and binary analysis as tools like BOLT identify thunks by disassembly.

smithp35 committed:
[LLD][AArch64][ARM] Delay adding long thunk mapping symbols

When we create a thunk we don't know whether it will be short or
long. Move the emission of the long thunk mapping symbol to
when we transition to a long thunk. This improves disassembly and
binary analysis as tools like BOLT identify thunks by disassembly.

smithp35 commented:
This removes the FIXME in the newly added aarch64-thunk-bti-multipass.s test. I've updated all the Arm thunks that have additional mapping symbols using the same pattern. I've not written any additional tests for all the Arm as these are quite difficult to write. I can add tests for some or all of these if necessary.

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -69,6 +69,7 @@ class AArch64Thunk : public Thunk {
 private:
   bool mayUseShortThunk = true;
   virtual void writeLong(uint8_t *buf) = 0;
+  virtual void addLongMapSyms() {}
```
Perhaps add some comment, something like: mapping symbols with non-zero offsets may become needed when mayUseShortThunk becomes false later.

smithp35 replied:
Thanks for the review, I've updated the comment. I'll also update the description as suggested below when merging.

MaskRay commented:
> This removes the FIXME in the newly added aarch64-thunk-bti-multipass.s test. I've updated all the Arm thunks that have additional mapping symbols using the same pattern. I've not written any additional tests for all the Arm as these are quite difficult to write. I can add tests for some or all of these if necessary.
> 
> I considered an alternative approach of adding a finalize() member function to ThunkSection and having that be called once per ThunkSection, that would then know for sure whether a thunk is short or long and could add the additional symbols. That would require another pass through all the input sections to find the ThunkSections though so unless we need finalize for something else I went with adding the additional symbols when we transition from short to long.

Thanks. The current approach looks good to me.

It may be worth mentioning #108989 and aarch64-thunk-bti-multipass.s in the description.

smithp35 committed:
Improve comments for addLongMapSyms

Explain why the function is needed. We don't know for certain at
creation time whether a long thunk will be written so delay the
choice until we know for certain.
